lib/srv/db/postgres/engine.go

```diff
@@ -340,6 +309,56 @@ func (e *Engine) receiveFromClient(client *pgproto3.Backend, server *pgproto3.Fr
 	}
 }
 
+// auditQueryMessage processes Query wire message which indicates that client
+// is executing a simple query.
+func (e *Engine) auditQueryMessage(session *common.Session, msg *pgproto3.Query) {
+	e.Audit.OnQuery(e.Context, session, common.Query{Query: msg.String})
+}
+
+// handleParseMesssage processes Parse wire message which indicates start of the
+// extended query protocol (prepared statements):
+// https://www.postgresql.org/docs/10/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY
+func (e *Engine) auditParseMessage(session *common.Session, msg *pgproto3.Parse) {
+	e.Audit.EmitEvent(e.Context, makeParseEvent(session, msg.Name, msg.Query))
+}
+
+// auditBindMessage processes Bind wire message which readies existing prepared
+// statement for execution into what Postgres calls a "destination portal",
+// optionally binding it with parameters (for parameterized queries).
+func (e *Engine) auditBindMessage(session *common.Session, msg *pgproto3.Bind) {
+	e.Audit.EmitEvent(e.Context, makeBindEvent(session, msg.PreparedStatement,
+		msg.DestinationPortal, formatParameters(msg.Parameters,
+			msg.ParameterFormatCodes)))
+}
+
+// auditExecuteMessage processes Execute wire message which indicates that
+// client is executing the previously parsed and bound prepared statement.
+func (e *Engine) auditExecuteMessage(session *common.Session, msg *pgproto3.Execute) {
+	e.Audit.EmitEvent(e.Context, makeExecuteEvent(session, msg.Portal))
+}
+
+// auditCloseMessage processes Close wire message which indicates that client
+// is closing a prepared statement or a destination portal.
+func (e *Engine) auditCloseMessage(session *common.Session, msg *pgproto3.Close) {
+	switch msg.ObjectType {
+	case closeTypePreparedStatement:
+		e.Audit.EmitEvent(e.Context, makeCloseEvent(session, msg.Name, ""))
+	case closeTypeDestinationPortal:
+		e.Audit.EmitEvent(e.Context, makeCloseEvent(session, "", msg.Name))
+	}
+}
+
+// auditFuncCallMessage processes FunctionCall wire message which indicates
+// that client is executing a system function.
+func (e *Engine) auditFuncCallMessage(session *common.Session, msg *pgproto3.FunctionCall) {
+	var formatCodes []int16
+	for _, fc := range msg.ArgFormatCodes {
+		formatCodes = append(formatCodes, int16(fc))
+	}
+	e.Audit.EmitEvent(e.Context, makeFuncCallEvent(session, int32(msg.Function),
```

Comment:
nit: `msg.Function` is uint32. assuming number is usually very small we should be ok, which should be always true. otherwise we may see negative number.

Reply:
Good catch, fixed.